rasterio/crs.py

```diff
@@ -1,111 +1,130 @@
 from rasterio.errors import CRSError
 from rasterio.compat import string_types
 
 
-def is_valid_crs(crs):
-    """Check if valid geographic or projected coordinate reference system."""
-    return is_geographic_crs(crs) or is_projected_crs(crs)
-
+class CRS(_CRS):
+    """A container class for coordinate reference system info
 
-def to_string(crs):
-    """Turn a parameter mapping into a more conventional PROJ.4 string.
+    PROJ.4 is the law of this land: http://proj.osgeo.org/. But whereas PROJ.4
+    coordinate reference systems are described by strings of parameters such as
 
-    Mapping keys are tested against the ``all_proj_keys`` list. Values of
-    ``True`` are omitted, leaving the key bare: {'no_defs': True} -> "+no_defs"
-    and items where the value is otherwise not a str, int, or float are
-    omitted.
-    """
-    items = []
-    for k, v in sorted(filter(
-            lambda x: x[0] in all_proj_keys and x[1] is not False and (
-                isinstance(x[1], (bool, int, float)) or
-                isinstance(x[1], string_types)),
-            crs.items())):
-        items.append(
-            "+" + "=".join(
-                map(str, filter(
-                    lambda y: (y or y == 0) and y is not True, (k, v)))))
-    return " ".join(items)
+        +proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs
 
+    here we use mappings:
 
-def from_string(prjs):
-    """Turn a PROJ.4 string into a mapping of parameters.
+        {'proj': 'longlat', 'ellps': 'WGS84', 'datum': 'WGS84', 'no_defs': True}
 
-    Bare parameters like "+no_defs" are given a value of ``True``. All keys
-    are checked against the ``all_proj_keys`` list.
+    One can set/get any PROJ.4 parameter using a dict-like key/value pair on the
+    object. You can instantiate the object by simply passing a dict to the
+    constructor. E.g.
 
-    EPSG:nnnn is allowed.
+        crs = CRS({'init': 'epsg:3005'})
 
-    JSON text-encoded strings are allowed.
     """
-    if '{' in prjs:
-        # may be json, try to decode it
-        try:
-            val = json.loads(prjs, strict=False)
-        except ValueError:
-            raise CRSError('crs appears to be JSON but is not valid')
-
-        if not val:
-            raise CRSError("crs is empty JSON")
-        else:
-            return val
-
-    if prjs.strip().upper().startswith('EPSG:'):
-        return from_epsg(prjs.split(':')[1])
-
-    parts = [o.lstrip('+') for o in prjs.strip().split()]
-
-    def parse(v):
-        if v in ('True', 'true'):
-            return True
-        elif v in ('False', 'false'):
-            return False
-        else:
-            try:
-                return int(v)
-            except ValueError:
-                pass
+
+    @property
+    def is_valid(self):
+        """Check if valid geographic or projected coordinate reference system."""
+        return self.is_geographic or self.is_projected
+
+    @property
+    def is_epsg_code(self):
+        for val in self.values():
```

Comment:
@brendan-ward agreed. By my count the whole line would be 80 chars exactly so let's just keep it as one line. As it says in PEP8 "However, know when to be inconsistent -- sometimes style guide recommendations just aren't applicable. When in doubt, use your best judgment"